shell-integration: switch to $GHOSTTY_SHELL_FEATURES

jparise opened the pull request:
This change consolidates all three opt-out shell integration environment variables into a single opt-in $GHOSTTY_SHELL_FEATURES variable. Its value is a comma-delimited list of the enabled shell feature names (e.g. "cursor,title").

$GHOSTTY_SHELL_FEATURES is set at runtime and automatically added to the shell environment. Its value is based on the shell-integration-features configuration option.

$GHOSTTY_SHELL_FEATURES is only set when at least one shell feature is enabled. It won't be set when 'shell-integration-features = false'.

$GHOSTTY_SHELL_FEATURES lists only the enabled shell feature names. We could have alternatively gone in the opposite direction and listed the disabled features, letting the scripts assume each feature is on by default like we did before, but I think this explicit approach is a little safer and easier to reason about / debug.

It also doesn't support the "no-" negation prefix used by the config system (e.g. "cursor,no-title"). This simplifies the implementation requirements of our (multiple) shell integration scripts, and because $GHOSTTY_SHELL_FEATURES is derived from shell-integration-features, the user-facing configuration interface retains that expressiveness.

$GHOSTTY_SHELL_FEATURES is intended to primarily be an internal concern: an interface between the runtime and our shell integration scripts. It could be used by people with particular use cases who want to manually source those scripts, but that isn't the intended audience.

... and because the previous $GHOSTTY_SHELL_INTEGRATION_NO_* variables were also meant to be an internal concern, this change does not include backwards compatibility support for those names.

One last advantage of a using a single $GHOSTTY_SHELL_FEATURES variable is that it can be easily forwarded to e.g. ssh sessions or other shell environments.

See: #5070

jparise committed:
shell-integration: switch to $GHOSTTY_SHELL_FEATURES

This change consolidates all three opt-out shell integration environment
variables into a single opt-in $GHOSTTY_SHELL_FEATURES variable. Its
value is a comma-delimited list of the enabled shell feature names (e.g.
"cursor,title").

$GHOSTTY_SHELL_FEATURES is set at runtime and automatically added to the
shell environment. Its value is based on the shell-integration-features
configuration option.

$GHOSTTY_SHELL_FEATURES is only set when at least one shell feature is
enabled. It won't be set when 'shell-integration-features = false'.

$GHOSTTY_SHELL_FEATURES lists only the enabled shell feature names. We
could have alternatively gone in the opposite direction and listed the
disabled features, letting the scripts assume each feature is on by
default like we did before, but I think this explicit approach is a
little safer and easier to reason about / debug.

It also doesn't support the "no-" negation prefix used by the config
system (e.g. "cursor,no-title"). This simplifies the implementation
requirements of our (multiple) shell integration scripts, and because
$GHOSTTY_SHELL_FEATURES is derived from shell-integration-features,
the user-facing configuration interface retains that expressiveness.

$GHOSTTY_SHELL_FEATURES is intended to primarily be an internal concern:
an interface between the runtime and our shell integration scripts. It
could be used by people with particular use cases who want to manually
source those scripts, but that isn't the intended audience.

... and because the previous $GHOSTTY_SHELL_INTEGRATION_NO_* variables
were also meant to be an internal concern, this change does not include
backwards compatibility support for those names.

One last advantage of a using a single $GHOSTTY_SHELL_FEATURES variable
is that it can be easily forwarded to e.g. ssh sessions or other shell
environments.

jparise commented on src/termio/shell_integration.zig:
```diff
@@ -150,9 +150,18 @@ pub fn setupFeatures(
-    if (!features.cursor) try env.put("GHOSTTY_SHELL_INTEGRATION_NO_CURSOR", "1");
-    if (!features.sudo) try env.put("GHOSTTY_SHELL_INTEGRATION_NO_SUDO", "1");
-    if (!features.title) try env.put("GHOSTTY_SHELL_INTEGRATION_NO_TITLE", "1");
+    var enabled = try std.BoundedArray(u8, 256).init(0);
```
I think we could also compute an precise array capacity at comptime.

pluiedev commented on src/shell-integration/fish/vendor_conf.d/ghostty-shell-integration.fish:
```diff
@@ -49,10 +49,10 @@ status --is-interactive || ghostty_exit
 
-    # Check if we are setting cursors
-    set --local no_cursor "$GHOSTTY_SHELL_INTEGRATION_NO_CURSOR"
+    set --local cursor string match -q "*cursor*" "$GHOSTTY_SHELL_FEATURES"
+    set --local sudo string match -q "*sudo*" "$GHOSTTY_SHELL_FEATURES"
```
For Fish at least I don't think this is the right approach. Fish has builtin list types and string operation primitives that guarantee the env var must be in a comma-delimited string. A combination of `string split` and `contains` makes much more sense IMO.

jparise replied:
That's a good idea. Thanks for the suggestion!

jparise committed:
shell-integration: use fish's native list type

Instead of looking for individual substrings in $GHOSTTY_SHELL_FEATURES,
`string split` it into a list of feature names and use `contains` to
detect their presence.

jparise committed:
shell-integration: use elvish's native list type

Instead of looking for individual substrings in $GHOSTTY_SHELL_FEATURES,
`str:split` it into a list of feature names and use `has-value` to
detect their presence.

jparise committed:
shell-integration: comptime buffer capacity

jparise committed:
shell-integration: minor documentation updates

mitchellh commented on src/termio/shell_integration.zig:
```diff
@@ -144,15 +144,29 @@ test "force shell" {
+            if (buffer.len > 0) try buffer.append(',');
+            try buffer.appendSlice(field.name);
+        }
+    }
```
I think it'd be a good idea to extract this into a function that we can unit test. (e.g.. set all features, set none, etc. etc.)

jparise replied:
@mitchellh just following up in case you didn't see my question

pluiedev reviewed (approved):
Looks good on my end, but should we have any sort of backwards compatibility? I get that this is supposed to be a private API but I wonder if people are doing something funny with these env vars

00-kat commented:
> I wonder if people are doing something funny with these env vars

I believe at least one person set them in a remote's shell config after copying the shell integration over.  Though I highly doubt there's much fishy (pun intended…) business going on.

mitchellh commented:
> ... and because the previous $GHOSTTY_SHELL_INTEGRATION_NO_* variables were also meant to be an internal concern, this change does not include backwards compatibility support for those names.

I agree with this, so I would say there's no backwards compatibility concerns. We document how to manually source the configs but that's still in the context of running it through Ghostty (just disabling auto _injection_). So this is good to go.

mitchellh reviewed (changes requested):
Just making sure I put a request changes in here for the unit test coverage.

jparise commented:
> > ... and because the previous $GHOSTTY_SHELL_INTEGRATION_NO_* variables were also meant to be an internal concern, this change does not include backwards compatibility support for those names.
> 
> I agree with this, so I would say there's no backwards compatibility concerns. We document how to manually source the configs but that's still in the context of running it through Ghostty (just disabling auto _injection_). So this is good to go.

The other relevant context is that we now always set these feature variables as of #5048. Prior to that, when shell integration itself was disabled, we wouldn't provide these feature variables, leaving that to the user for the manual sourcing case.

mitchellh reviewed (approved):
Sorry for the delay!
